File: termux-display/src/main/cpp/xserver/hw/xnest/Visual.c
/*

Copyright 1993 by Davor Matic

Permission to use, copy, modify, distribute, and sell this software
and its documentation for any purpose is hereby granted without fee,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation.  Davor Matic makes no representations about
the suitability of this software for any purpose.  It is provided "as
is" without express or implied warranty.

*/

#ifdef HAVE_XNEST_CONFIG_H
#include <xnest-config.h>
#endif

#include <X11/X.h>
#include <X11/Xproto.h>
#include "scrnintstr.h"
#include "dix.h"
#include "mi.h"
#include "Xnest.h"

#include "Display.h"
#include "Visual.h"

Visual *
xnestVisual(VisualPtr pVisual)
{
    int i;

    for (i = 0; i < xnestNumVisuals; i++)
        if (pVisual->class == xnestVisuals[i].class &&
            pVisual->bitsPerRGBValue == xnestVisuals[i].bits_per_rgb &&
            pVisual->ColormapEntries == xnestVisuals[i].colormap_size &&
            pVisual->nplanes == xnestVisuals[i].depth &&
            pVisual->redMask == xnestVisuals[i].red_mask &&
            pVisual->greenMask == xnestVisuals[i].green_mask &&
            pVisual->blueMask == xnestVisuals[i].blue_mask)
            return xnestVisuals[i].visual;

    return NULL;
}

Visual *
xnestVisualFromID(ScreenPtr pScreen, VisualID visual)
{
    int i;

    for (i = 0; i < pScreen->numVisuals; i++)
        if (pScreen->visuals[i].vid == visual)
            return xnestVisual(&pScreen->visuals[i]);

    return NULL;
}

Colormap
xnestDefaultVisualColormap(Visual * visual)
{
    int i;

    for (i = 0; i < xnestNumVisuals; i++)
        if (xnestVisuals[i].visual == visual)
            return xnestDefaultColormaps[i];

    return None;
}

File: termux-display/src/main/java/com/termux/display/controller/xserver/Visual.java
package com.termux.display.controller.xserver;

public class Visual {
    public final int id;
    public final byte visualClass = 4;
    public final boolean displayable;
    public final byte depth;
    public final byte bitsPerRGBValue;
    public final short colormapEntries = 256;
    public final int blueMask;
    public final int greenMask;
    public final int redMask;

    public Visual(int id, boolean displayable, int depth, int bitsPerRGBValue, int redMask, int greenMask, int blueMask) {
        this.id = id;
        this.displayable = displayable;
        this.depth = (byte)depth;
        this.bitsPerRGBValue = (byte)bitsPerRGBValue;
        this.redMask = redMask;
        this.greenMask = greenMask;
        this.blueMask = blueMask;
    }
}
